app/services/activities/import_from_csv.rb

```diff
@@ -61,6 +74,54 @@ def update
           @errors[attr_name] ||= [@converter.raw(attr_name), message]
         end
       end
+
+      def find_activity_by_roda_id(roda_id)
+        activity = Activity.by_roda_identifier(roda_id)
+        @errors[:roda_id] ||= [roda_id, I18n.t("importer.errors.activity.not_found")] if activity.nil?
+
+        activity
+      end
+    end
+
+    class ActivityCreator
```

Comment:
This class does exactly what I was expecting in terms of using of identifiers: passing `Parent RODA ID` to `Activity.by_roda_identifier` to get the parent, assigning `roda_identifier_fragment` and using `cache_roda_identifier` to generate the full ID 👍 

A restriction implied in this design is that a CSV import can only create activities whose parent is already in the system, and that parent must have a RODA ID assigned. That sidesteps a bunch of problems to do with ordering the import rows -- to assign a RODA ID to an activity, its parent must already have an ID, so these IDs must be assigned in activity-level order.

By using the RODA ID to identify the parent, that means this system can only create activities where assigning a RODA ID is safe, because if the parent exists then it must have a RODA ID already.

Not sure if I've explained that very well, the point is this avoids a bunch of nasty problems and I like it :)